nemo/collections/asr/parts/k2/w_transducer.py

```diff
@@ -0,0 +1,337 @@
+            .flatten()
+        )
+        to_states = from_states + 1
+        ilabels = units_tensor.expand(num_frames, -1).flatten()
+        arcs[num_forward_arcs:-2, 0] = from_states
+        arcs[num_forward_arcs:-2, 1] = to_states
+        arcs[num_forward_arcs:-2, 2] = ilabels
+
+        # last 2 states
+        arcs[-2, :3] = torch.tensor((num_grid_states - 1, num_grid_states, blank_id), dtype=torch.int32, device=device)
+        arcs[-1, :3] = torch.tensor((num_grid_states, num_grid_states + 1, -1), dtype=torch.int32, device=device)
+
+        # sequence indices, time indices
+        olabels = torch.div(arcs[:, 0], (text_length + 1), rounding_mode="floor")  # arcs[:, 0] // (text_length + 1)
+        unit_positions = arcs[:, 0] % (text_length + 1)
+        # last state: final
+        olabels[-1] = -1
+        unit_positions[-1] = -1
+
+        # relabel
+        # instead of using top sort (extremely expensive) k2.top_sort(rnnt_graph)
+        arcs[:-2, 0] = self.relabel_states(arcs[:-2, 0], text_length + 1, num_frames)
+        arcs[:-3, 1] = self.relabel_states(arcs[:-3, 1], text_length + 1, num_frames)
+
+        if self.last_blank_mode == self.LastBlankMode.ALLOW_IGNORE:
+            arcs[
+                num_forward_arcs_base + (num_frames - 1) : num_forward_arcs_base + (num_frames - 1) * 2, 1
+            ] = num_grid_states
+
+        # sort by start state - required in k2
+        # TODO: maybe it is more optimal to avoid sort, construct arcs in ascending order
+        _, indices = torch.sort(arcs[:, 0], dim=0)
+        arcs = arcs[indices]
+        olabels = olabels[indices]
+        unit_positions = unit_positions[indices]
+
+        rnnt_graph = k2.Fsa(arcs, olabels)
+        rnnt_graph.unit_positions = unit_positions
+        return rnnt_graph
+
+    def forward(
+        self, acts: torch.Tensor, labels: torch.Tensor, act_lens: torch.Tensor, label_lens: torch.Tensor,
+    ):
+        """
+        Forward method is similar to RNN-T Graph-Transducer forward method,
+        but we need to assign eps weight to eps-transitions.
+        """
+        # argument names are consistent with NeMo, see RNNTLoss.forward:
+        # self._loss(acts=log_probs, labels=targets, act_lens=input_lengths, label_lens=target_lengths)
+        logits, targets, logits_lengths, target_lengths = acts, labels, act_lens, label_lens
+
+        # logits: B x Time x Text+1 x C
+        vocab_size = logits.shape[-1]
+        target_fsas_vec = self.get_graphs_batched(logits_lengths, targets, target_lengths, vocab_size)
+
+        cast_context = force_float32_context() if self.cast_to_float32 else nullcontext()
+        with cast_context:
+            log_probs = F.log_softmax(logits, dim=-1)
+            with torch.no_grad():
+                indices = self.get_logits_indices(target_fsas_vec, logits.shape)
+                indices[target_fsas_vec.labels == -1] = 0
+                indices[target_fsas_vec.labels >= vocab_size] = 0  # eps
```

Comment:
Oh, I see. Well, anyway, it's better to address such things before `forward` and presume that the lattices are without redundancies at this point.

Reply:
Added comments, as we discussed